openmc/checkvalue.py

```diff
@@ -0,0 +1,337 @@
+            msg = 'Unable to set "{0}" to "{1}" since it is greater than ' \
+                  '"{2}"'.format(name, value, maximum)
+            raise ValueError(msg)
+    else:
+        if value >= maximum:
+            msg = 'Unable to set "{0}" to "{1}" since it is greater than ' \
+                  'or equal to "{2}"'.format(name, value, maximum)
+            raise ValueError(msg)
+
+def check_greater_than(name, value, minimum, equality=False):
+    """Ensure that an object's value is greater than a given value.
+
+    Parameters
+    ----------
+    name : str
+        Description of the value being checked
+    value : object
+        Object to check
+    minimum : object
+        Minimum value to check against
+    equality : bool, optional
+        Whether equality is allowed. Defaults to False.
+
+    """
+
+    if equality:
+        if value < minimum:
+            msg = 'Unable to set "{0}" to "{1}" since it is less than ' \
+                  '"{2}"'.format(name, value, minimum)
+            raise ValueError(msg)
+    else:
+        if value <= minimum:
+            msg = 'Unable to set "{0}" to "{1}" since it is less than ' \
+                  'or equal to "{2}"'.format(name, value, minimum)
+            raise ValueError(msg)
+
+
+def check_filetype_version(obj, expected_type, expected_version):
+    """Check filetype and version of an HDF5 file.
+
+    Parameters
+    ----------
+    obj : h5py.File
+        HDF5 file to check
+    expected_type : str
+        Expected file type, e.g. 'statepoint'
+    expected_version : int
+        Expected major version number.
+
+    """
+    try:
+        this_filetype = obj.attrs['filetype'].decode()
+        this_version = obj.attrs['version']
+
+        # Check filetype
+        if this_filetype != expected_type:
+            raise IOError('{} is not a {} file.'.format(
+                obj.filename, expected_type))
+
+        # Check version
+        if this_version[0] != expected_version:
+            raise IOError('{} file has a version of {} which is not '
+                          'consistent with the version expected by OpenMC, {}'
+                          .format(this_filetype, 
+                              '.'.join(str(v) for v in this_version),
+                              expected_version))
+    except AttributeError:
+        raise IOError('Could not read {} file. This most likely means the {} '
+                      'file was produced by a different version of OpenMC than '
+                      'the one you are using.'.format(expected_type))
+
+
+class CheckedList(list):
+    """A list for which each element is type-checked as it's added
+
+    Parameters
+    ----------
+    expected_type : type or Iterable of type
+        Type(s) which each element should be
+    name : str
+        Name of data being checked
+    items : Iterable, optional
+        Items to initialize the list with
+
+    """
+
+    def __init__(self, expected_type, name, items=[]):
```

Comment:
Using an empty list (or any mutable argument) as a default argument is a well-known source of errors and is highly recommended against. Most IDEs will complain about this with the "default argument value is mutable" error. 

The solution to this is to pass `None` as the default argument and set the value to `[]` inside the code.

Here this amounts to,

```
def __init__(self, expected_type, name, items=None):
    if items is None:
        items = []
    # rest of code
```

For reference/reading on this issue
1. https://docs.quantifiedcode.com/python-anti-patterns/correctness/mutable_default_value_as_argument.html
2. https://docs.python-guide.org/writing/gotchas/
3. https://stackoverflow.com/questions/101268/hidden-features-of-python#113198